# Tools making optimisation problem

AAA is a company that produces tools T1, T2 and T3. These are are made from steel, and the process involves molding the tools and then assembling them.

The following table summarises the requirements and constraints:

|                          | T1     | T2     | T3     | Availability |
|--------------------------|--------|--------|--------|--------------|
| Steel (kg)               | 1      | 0.7    | 0.6    | 10,000       |
| Molding machine (hr)     | 1      | 1      | 1      | 20,000       |
| Assembly machine (hr)    | 0.3    | 0.5    | 0.4    | 9,000        |
| Demand limit (tools/day) | 15,000 | 16,000 | 12,000 |              |
| Profit (£/1000 units)    | 130    | 100    | 120    |              |

AAA would like to plan for the daily production of tools to maximise its profit.
In particular, we have the following questions:

- How many of each type of the tools should AAA plan to produce per day?
- What is the expected total profit from this plan?
- Which resources are expected to be most critical in this plan?

### Marking scheme

|Item|Mark|
|:----|---:|
|**Mathematical formulation**:||
|Decision variables|/1|
|Profit formula|/1|
|Constraints|/2|
|Demand limits|/2|
|Non-negativity constraints|/1|
|**Solution**:||
|PuLP solution|/4|
|**Answers to AAA's questions**:||
|How many of each type of the tools should AAA plan to produce per day?|/3|
|What is the expected total profit from this plan?|/3|
|Which resources are expected to be most critical in this plan?|/3|
|||
|**Total**:     |/20|


## Mathematical formulation

Decision variables:

|Variable name| Description |
|A1 |This will represent T1 |

|A2|This will represent T2|

|A3|This will represent T3|

|model_X|This will represent profit|


Profit formula (to be optimised):
$$
....
$$

In [ ]:
model_X += 130*A1 + 100*A2 +120*A3 

Constraints:

$$
\begin{alignat*}{4}
    &1*A1 + 0.7*A2 + 0.6*A3 <=10,000& \leq \\

&1*A1 + 1*A2 + 1*A3 <=20,000& \leq \\

& 0.3*A1 + 0.5*A2 + 0.4*A3 <=9,000& \leq \\ 
    
\end{alignat*}
$$

Demand limits:

$$
\begin{alignat*}{2}
    &A1<= 15000 &\leq .. \\
    &A2<= 16000 &\leq .. \\
    &A3<= 12000&\leq .. \\
\end{alignat*}
$$

Non-negativity constraints:

$$
 A1,A2,A3,model_X >=0
$$

## Solution

In [21]:
import pulp 
from pulp import LpMaximize,LpProblem , LpVariable, LpInteger,LpStatus

In [22]:
#Creating model_X variable to contain the problem data 
model_X = pulp.LpProblem("Profit maximising",pulp.LpMaximize)
#defining the variables T1 T2 and T3 which are created with the lower limit or lower bound 0
A1 = pulp.LpVariable("T1",0 )
A2 = pulp.LpVariable("T2",0)
A3 = pulp.LpVariable("T3",0 )

C:\Users\sneha\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\pulp\pulp.py:1352: UserWarning: Spaces are not permitted in the name. Converted to '_'
  warnings.warn("Spaces are not permitted in the name. Converted to '_'")


In [23]:
#Maximzie profite by creating the objective function
model_X += 130*A1 + 100*A2 +120*A3 ,"Profit"


In [24]:
# creating the constrains 
model_X +=1*A1 + 0.7*A2 + 0.6*A3 <=10,000
model_X +=1*A1 + 1*A2 + 1*A3 <=20,000
model_X +=0.3*A1 + 0.5*A2 + 0.4*A3 <=9,000

In [25]:
#seting a demand limit
model_X +=A1<=15,000
model_X +=A2<=16,000
model_X +=A3<=12,000

In [26]:
#non negative constrains:
A1 ,A2 , A3 >=0

(T1, T2, 1*T3 + 0 >= 0)

In [27]:
# Solve our problem
model_X.solve()
pulp.LpStatus[model_X.status]

'Optimal'

In [28]:

#PRINTING OUR DECISION VARIABLE
print (f"Tool T1 = {format(A1.varValue)}\n")
print (f"Tool T2 = {format(A2.varValue)}\n")
print (f"Tool T3 = {format(A3.varValue)}\n")

Tool T1 = 0.0

Tool T2 = 4.0

Tool T3 = 12.0



In [29]:
# Print our objective function value
print (pulp.value(model_X.objective))

1840.0


In [30]:
#Daily resource score 
print("resource value")
print ("steel",A1.varValue*1+A2.varValue*0.7+A3.varValue*0.6)
print('molding machine',A1.varValue*1+A2.varValue*1+A3.varValue*1)
print("assembly machine",A1.varValue*0.3+A2.varValue*0.5+A3.varValue*0.4 )
print ("production time",A1.varValue*1+A2.varValue*1+A3.varValue*1+A1.varValue*0.3+A2.varValue*0.5+A3.varValue*0.4)

resource value
steel 10.0
molding machine 16.0
assembly machine 6.800000000000001
production time 22.8


# Conclusion

How many of each type of the tools should AAA plan to produce per day?
Tool T1 = 0.0 or 000
Tool T2 = 4.0 or 4000
Tool T3 = 12.0 or 12000

What is the expected total profit from this plan?

1840.0 

Which resources are expected to be most critical in this plan?
resource value
steel 10.0 or 10000
molding machine 16.0 or 16000
assembly machine 6.800000000000001 or 68000
production time 22.8 or 22800

when we take a look at the constrains we can observe that the steel has reached the availability when it will decrease or increase it may effect the construction cost so we can consider it a most critical in the plan 



# List of references

Ben Keen(1st April 2016)Linear Programming with Python and PuLP https://benalexkeen.com/linear-programming-with-python-and-pulp-part-3/

pulp documentation team Linear Programming: https://coin-or.github.io/pulp/